game/groundforces/template.py

```diff
@@ -0,0 +1,442 @@
+from __future__ import annotations
+
+import copy
+import random
+from dataclasses import dataclass, field
+from typing import Iterator, Any, TYPE_CHECKING, Optional
+
+from dcs import Point
+from dcs.unit import Unit
+from dcs.unittype import VehicleType, ShipType
+from game.data.radar_db import UNITS_WITH_RADAR
+from game.data.units import UnitClass
+from game.data.groups import GroupRole, GroupTask, ROLE_TASKINGS
+
+from game.dcs.groundunittype import GroundUnitType
+from game.dcs.shipunittype import ShipUnitType
+from game.dcs.unittype import UnitType
+from game.theater.theatergroundobject import (
+    SamGroundObject,
+    EwrGroundObject,
+    BuildingGroundObject,
+    GroundGroup,
+    MissileSiteGroundObject,
+    ShipGroundObject,
+    CarrierGroundObject,
+    LhaGroundObject,
+    CoastalSiteGroundObject,
+    VehicleGroupGroundObject,
+    IadsGroundObject,
+    GroundUnit,
+)
+from game.point_with_heading import PointWithHeading
+
+from game.utils import Heading
+from game import db
+
+if TYPE_CHECKING:
+    from game import Game
+    from game.factions.faction import Faction
+    from game.theater import TheaterGroundObject, ControlPoint
+
+
+@dataclass
+class UnitTemplate:
+    name: str
+    position: Point
+    heading: int
+
+    @staticmethod
+    def from_unit(unit: Unit) -> UnitTemplate:
+        return UnitTemplate(
+            unit.name,
+            Point(int(unit.position.x), int(unit.position.y)),
+            int(unit.heading),
+        )
+
+
+@dataclass
+class GroupTemplate:
+    name: str
+    units: list[UnitTemplate]
+
+    # Is Static group
+    static: bool = False
+
+    # The group this template will be merged into
+    group: int = 1
+
+    # Define the amount of random units to be created by the randomizer.
+    # This can be a fixed int or a random value from a range of two ints as tuple
+    unit_count: list[int] = field(default_factory=list)
+
+    # defintion which unit types are supported
+    unit_types: list[str] = field(default_factory=list)
+    unit_classes: list[UnitClass] = field(default_factory=list)
+    alternative_classes: list[UnitClass] = field(default_factory=list)
+
+    # Defines if this groupTemplate is required or not
+    optional: bool = False
+
+    # Used to determine if the group should be generated or not
+    _possible_types: list[str] = field(default_factory=list)
+    _enabled: bool = True
+    _unit_type: Optional[str] = None
+    _unit_counter: Optional[int] = None
+
+    def initialize_for_faction(
+        self, faction: Faction, faction_sensitive: bool = True
+    ) -> bool:
```

Comment:
Same problem here as in `GroundForceGroup`, I think. There's only one copy of each named template, so applying faction behavior to it will alter the template for both red and blue.

Same fix too: don't modify the template. Instead, apply the faction specific behavior when the template is used.

Reply:
I thought the same before but this means that whenever a group will be generated we have to compute the usable units / faction specific behaviour. To prevent this redundant computation i've gone the way with deepcopys. Not quite sure if there is a "best" way to go. I personally do like the approach of calculate it once and reuse it many times more than write it generic and calculate many times. Debatable :)
If you wish i'll move away from the deepcopy. What do you say @DanAlbert ?